### PR TITLE
Added template configuration to task template

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
This PR adds a template field to the `TaskTemplate` which contains either a `PodTemplate` name or a marshaled `PodTemplate` struct.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3123

## Follow-up issue
_NA_

### CODE REVIEW

File: protos/flyteidl/core/tasks.proto [hunk 1]
```diff
@@ -143,6 +143,15 @@ message TaskTemplate {
     // to use as required.
     // reserve the field numbers 1 through 15 for very frequently occurring message elements
     map<string, string> config = 16;
+
+    // pod_template provides the definition for a PodTemplate to be applied as the base configuration for this task.
+    // This enables fine-grained k8s configuration for Flyte tasks in addition to overriding the default PodTemplate.
+    oneof pod_template {
+        // pod_template_name is the unique name of a PodTemplate k8s resource.
+        string pod_template_name = 19;
+        // pod_template_struct is a JSON marshaled PodTemplate.
```
Comment (EngHabu) [inline protos/flyteidl/core/tasks.proto:152]: Can you elaborate here on how overriding properties will work? Will service account be copied from the task instead.. how about container images... Etc. Etc.

**Review by EngHabu — APPROVED** [review]
Let's add more code comments to explain the behavior. I'm pretty sure we will get a lot of questions here...

File: protos/flyteidl/core/tasks.proto (hunk 1 above)
Comment (hamersaw) [reply protos/flyteidl/core/tasks.proto:152]: So we have put together pretty [extensive docs on configuring Flyte pods](https://docs.flyte.org/en/latest/deployment/cluster_config/general.html#configuring-k8s-pod) which highlights using the default PodTemplate vs k8s plugin configuration. I imagined we might extend on this (maybe make a separate page?) to try and sort this out end to end. How in-depth do you think is necessary here? We could fill 100 lines I imagine.

**Comment by kumare3** [comment]
Why do we need this, why not use the one of

**Comment by flixr** [comment]
@kumare3 e.g. for my usecases it boils down to setting some things like `runtimeClassName`, `affinity`, etc. per task (some tasks in the workflow need to run on nodes with large GPU, same only need a small one). And it would allow mounting of secrets, etc.. see also https://github.com/flyteorg/flyte/issues/3055

**Review by eapolinario — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]

**Pull request merged**